### PR TITLE
Feat (core/float): better max mantissa computation

### PR DESCRIPTION
**Pull request by Giuseppe5**
## Reason for this PR

Small refactor to improve computation of max available mantissa given a certain mantissa bit width.
This avoids data-dependent for loop, which has two main benefits:
- Allows for gradient flow of `mantissa_bit_width` through the function
- Allows for fullgraph compile through the quantizers

The last point seems particularly relevant. From a custom training script, the following times have been observed:
- Baseline - 700 imgs/s
- Compile before this change (with graph breaks) - 1400 imgs/s
- Compile with this change (no graph breaks) - 2000 imgs/s

## Risk Highlight

- [ ] This PR includes code from another work (please detail).
- [ ] This PR contains API-breaking changes.
- [ ] This PR depends on work in another PR (please provide links/details).
- [ ] This PR introduces new dependencies (please detail).
- [ ] There are coverage gaps not covered by tests.
- [ ] Documentation updates required in subsequent PR.

## Checklist

- [ ] Code comments added to any hard-to-understand areas, if applicable.
- [ ] Changes generate no new warnings.
- [ ] Updated any relevant tests, if applicable.
- [ ] No conflicts with destination `dev` branch.
- [ ] I reviewed my own code changes.
- [ ] Initial CI/CD passing.
- [ ] 1+ reviews given, and any review issues addressed and approved.
- [ ] Post-review full CI/CD passing.

### CODE REVIEW

File: tests/brevitas/core/test_float_quant.py [hunk 1]
```diff
@@ -31,7 +31,7 @@ class BitwidthWrapper(torch.nn.Module):
 
     def __init__(self, value):
         super().__init__()
-        self.value = torch.tensor(value)
+        self.value = torch.tensor(float(value))
```
Comment (nickfraser) [inline tests/brevitas/core/test_float_quant.py:36]: Not obvious why this change is necessary. Worth a comment?

File: tests/brevitas/core/test_float_quant.py [hunk 2]
```diff
@@ -252,7 +252,7 @@ def test_inner_scale(inp, minifloat_format, scale):
 
         # scale inp manually
         scaled_inp = inp / scale
-        max_mantissa = compute_max_mantissa(torch.tensor(mantissa_bit_width))
+        max_mantissa = compute_max_mantissa(torch.tensor(float(mantissa_bit_width)))
```
Comment (nickfraser) [inline tests/brevitas/core/test_float_quant.py:257]: Same here - not obvious why this change is necessary. Worth a comment?

**Comment by nickfraser** [comment]
Not sure why your tests are failing - I even sanity checked that your fix makes sense on my end:

```python
import torch

def cmm1(mm):
    return torch.sum((2. ** torch.arange(0, -1. * mm - 1., -1.)))

def cmm2(mm):
    return 2 * (1 - 2 ** (-mm - 1))

for i in range(0, 20):
    mm = float(i)
    assert cmm1(mm) == cmm2(mm)
```

^Works with no error.

**Review by nickfraser — APPROVED** [review]
A few comments. Not sure why this change is causing failing tests - it seems pretty benign to me!

Please fix the tests though 🙏

File: tests/brevitas/core/test_float_quant.py (hunk 1 above)
Comment (Giuseppe5) [reply tests/brevitas/core/test_float_quant.py:36]: And I had to add another one.

We need to make sure bitwidth is a float, so that the max mantissa computation is a float, otherwise it gets rounded to an int.

Comment (nickfraser) [reply tests/brevitas/core/test_float_quant.py:36]: Got it, but I meant a code comment ;)
